Debug Tests: Improve connection test, and allow overrides

roccotripaldi opened the pull request:
## Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Allowed failing and passing tests to override the proposed defaults for the Site Health, while preserving the functionality of test messaging in the CLI and API test environments.
* improved the messaging on the Connection test for the Site Health page.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This is step one in improving Jetpack's site health section
* See: p6TEKc-3st-p2

#### Testing instructions:
* Test that CLI tests still function
* Test that all Jetpack Site Health tests still look okay ( Tools -> Site Health )
* Take a look a the new Site Health card for a connected site:

<img width="882" alt="Screen Shot 2020-02-20 at 3 06 13 PM" src="https://user-images.githubusercontent.com/2694219/74975299-d80bd080-53f4-11ea-887b-d8acdbb3bc81.png">

* Take a look at the new Site Health card for an unconnected site:

<img width="850" alt="Screen Shot 2020-02-20 at 2 57 42 PM" src="https://user-images.githubusercontent.com/2694219/74975362-f1148180-53f4-11ea-98fc-93b3cc9e1b93.png">


#### Proposed changelog entry for your changes:
* n/a

roccotripaldi committed:
Jetpack CXN tests:

- Allow successful tests to provide their own message and label for the Site Health page.
- Adjust the message and label for the `test__check_if_connected` test

roccotripaldi committed:
Debug tests

- Adds the ability for tests to supply an action label and override other labels.
- Improves the failing state for `test__check_if_connected`

kraftbj commented on _inc/lib/debugger/class-jetpack-cxn-test-base.php:
```diff
@@ -282,12 +294,17 @@ public static function skipped_test( $name = 'Unnamed', $message = false ) {
 		// Provide standard resolutions steps, but allow pass-through of non-standard ones.
 		switch ( $resolution ) {
+			case 'connect_jetpack':
+				$resolution = false;
```
The resolution field is for a human-readable action. If the failing result is passing a custom one, it doesn't need to be listed here since it'll just pass through (L324 as modified).

kraftbj commented on _inc/lib/debugger/class-jetpack-cxn-tests.php:
```diff
@@ -91,11 +91,30 @@ public static function increase_timeout() {
 		if ( $this->helper_is_jetpack_connected() ) {
-			$result = self::passing_test( $name );
+			$message = sprintf(
+				'<p>%s</p><p>%s</p>',
```
May need to check this to make sure the output works fine with CLI.

kraftbj replied:
What we could do is, when outputting the error message for the Site Health section, run the message through wpautop() to add paragraph tags to the text. The raw error message (and what the API and CLI output) is plain text, but add the HTML for Site Health?

roccotripaldi committed:
Debug tests

Allow passing tests to override defaults with an HTML description.

roccotripaldi committed:
Debug tests

- perserve plain text messages to use on API tests
- allow tests to override plain text with HTML for the Site Health page.

roccotripaldi commented:
Thanks for the review @kraftbj  - i had no idea that these tests were also used in CLI. I've made changes to preserve CLI functionality while allowing overrides for HTML in Site Health

jeherve commented on _inc/lib/debugger/class-jetpack-cxn-tests.php:
```diff
@@ -91,11 +91,34 @@ public static function increase_timeout() {
+				sprintf(
+					'<p>%s</p><p>%s</p>',
+					__( 'A healthy connection ensures Jetpack essential services are provided to your WordPress site, such as Stats and Site Security.', 'jetpack' ),
+					__( '✅ Your site is connected to Jetpack.', 'jetpack' )
```
Maybe we should pull the emoji out of the translation and use a placeholder instead? I am not sure translators / translator tools can all handle adding emoji.

keoshi replied:
Agreed — ideally this should reuse a Dashicon instead of being an emoji:

Green/Healthy/Good:

```
<span class="dashicons pass"><span class="screen-reader-text">Passed</span></span>
```

Read/Error/Bad:

```
<span class="dashicons fail"><span class="screen-reader-text">Error</span></span>
```

jeherve commented on _inc/lib/debugger/class-jetpack-cxn-tests.php:
```diff
@@ -91,11 +91,34 @@ public static function increase_timeout() {
+				__( 'Your site is not connected to Jetpack', 'jetpack' ),
+				__( 'Learn more about this process', 'jetpack' ),
+				sprintf(
+					'<p>%s</p><p>%s<strong>%s</strong></p>',
```
Just a nit-pick, but I personally find it easier to read when we use numbered placeholders when we have more than a couple of them. Up to you!

jeherve commented on _inc/lib/debugger/class-jetpack-cxn-test-base.php:
```diff
@@ -277,15 +291,18 @@ public static function skipped_test( $name = 'Unnamed', $message = false ) {
 	 * @return array Test results.
 	 */
-	public static function failing_test( $name, $message, $resolution = false, $action = false, $severity = 'critical' ) {
+	public static function failing_test( $name, $message, $resolution = false, $action = false, $severity = 'critical', $label = false, $action_label = 'Resolve', $description = false ) {
```
Maybe we can pull `Resolve` out of there so it can still be translated?

```suggestion
	public static function failing_test( $name, $message, $resolution = false, $action = false, $severity = 'critical', $label = false, $action_label = '', $description = false ) {
		if ( empty( $action_label ) ) {
			$action_label = __( 'Resolve', 'jetpack' );
		}
```

keoshi reviewed (commented):
Looks good to me! I've changed the failed state a tiny bit in iteration no. 2 though, which will be posted shortly.

keoshi commented:
Designs i2 posted p6TEKc-3t2-p2

roccotripaldi committed:
Debug tests
- adds dashicons instead of emoji
- improves translation logic

jeherve reviewed (dismissed):
It tests well for me. 👍

kraftbj commented:
A couple of comments:

1. Seeing the function args grow and grow, it may make sense to migrate to passing an args array.

2. Breaking out the HTML description from the plain text one is giving me pause. To me, it feels like a case where we would possibly be duplicating a lot of text. Using wpautop, could we do this with an extra field? Links are the one thing I'm less sure about though we can make them clickable through the filter that does that in Core already. 

Having multiple strings show up in different places could complicate seeking support or answers if it isn't clear that multiple messages are referring to the same thing.

mdbitz committed:
[not verified] Updated Action to  per i2 designs, learn more will be re-added as part of next sprint when multiple action links are supported

mdbitz commented:
Per the i2 designs the primary action text should be "Reconnect your site now" for failed connections.

mdbitz commented:
Note about changelog we could use something around the lines of:

Site Health Tests
Updated Connect test with detailed descriptions and actions to resolve failing tests.
